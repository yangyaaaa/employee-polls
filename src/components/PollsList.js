import { connect } from "react-redux";
import PollItem from "./PollItem";
import { useState, useEffect } from "react";

const PollsList = (props) => {
  const [polls, setPolls] = useState(props.unAnsweredPolls);

  useEffect(() => {
    const userAddedQuestions = props.userQuestions.filter(
      (quetion) =>
        !props.unAnsweredPolls.includes(quetion) &&
        !props.answeredPolls.includes(quetion)
    );
    const defaultList = props.unAnsweredPolls.concat(userAddedQuestions);
    setPolls(defaultList);
  }, [props.userQuestions]);

  const showAnsweredPolls = () => {
    setPolls(props.answeredPolls);
  };

  const showUnAnsweredPolls = () => {
    setPolls(props.unAnsweredPolls);
  };

  return (
    <div className="poll-list-container">
      <div >
        <button
          
          onClick={showAnsweredPolls}
          data-testid="answered-polls-test-id"
        >
          Answered Polls
        </button>
        <button
          
          onClick={showUnAnsweredPolls}
          data-testid="unanswered-polls-test-id"
        >
          UnAnswered Polls
        </button>
      </div>

      <div >
        <ul>
          {polls.map((pollId) => (
            <li key={pollId} >
              <PollItem pollId={pollId} />
            </li>
          ))}
        </ul>
      </div>
    </div>
  );
};

const mapStateToProps = ({ polls, authedUser, users }) => {
  const user = users[authedUser];

  let authedUserAnsweredPolls = [];
  if (user && user.answers) {
    authedUserAnsweredPolls = Object.keys(user.answers);
  }

  const answeredPolls = Object.keys(polls)
    .filter((pollId) => authedUserAnsweredPolls.includes(pollId))
    .sort((a, b) => {
      return polls[b].timestamp - polls[a].timestamp;
    });

  const unAnsweredPolls = Object.keys(polls)
    .filter((pollId) => !authedUserAnsweredPolls.includes(pollId))
    .sort((a, b) => {
      return polls[b].timestamp - polls[a].timestamp;
    });

  let userQuestions = [];
  if (user) {
    userQuestions = user.questions;
  }

  return {
    answeredPolls,
    unAnsweredPolls,
    userQuestions
  };
};

export default connect(mapStateToProps)(PollsList);