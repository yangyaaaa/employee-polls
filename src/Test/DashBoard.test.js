import PollsList from "../components/PollsList";
import { render, screen } from "@testing-library/react";
import * as React from "react";

import { MemoryRouter } from "react-router";
import { Provider } from "react-redux";
import thunk from 'redux-thunk';
import configureStore from "redux-mock-store";


const mockedStore = configureStore([thunk]);

describe("DashBoard", () => {
  let mockStore;

  beforeEach(() => {
    mockStore = mockedStore(appData)});

  it("will catch a snapshot for DashBoard", () => {
    const view = render(
      <MemoryRouter>
        <Provider store={mockStore}>
          <PollsList />
        </Provider>
      </MemoryRouter>
    );
    expect(view).toMatchSnapshot();
  });

  it("will have expected elements", () => {
    render(
      <MemoryRouter>
        <Provider store={mockStore}>
          <PollsList />
        </Provider>
      </MemoryRouter>
    );
    const answeredTab = screen.getByTestId("answered-polls-test-id");
    expect(answeredTab).toBeInTheDocument();

    const unAnsweredTab = screen.getByTestId("unanswered-polls-test-id");
    expect(unAnsweredTab).toBeInTheDocument();
  });
});






const appData = {
  "authedUser": "sarahedo",
  "users": {
      "sarahedo": {
          "id": "sarahedo",
          "password": "password123",
          "name": "Sarah Edo",
          "avatarURL": null,
          "answers": {
              "8xf0y6ziyjabvozdd253nd": "optionOne",
              "6ni6ok3ym7mf1p33lnez": "optionOne",
              "am8ehyc8byjqgar0jgpub9": "optionTwo",
              "loxhs1bqm25b708cmbf3g": "optionTwo"
          },
          "questions": [
              "8xf0y6ziyjabvozdd253nd",
              "am8ehyc8byjqgar0jgpub9"
          ]
      },
      "tylermcginnis": {
          "id": "tylermcginnis",
          "password": "abc321",
          "name": "Tyler McGinnis",
          "avatarURL": null,
          "answers": {
              "vthrdm985a262al8qx3do": "optionOne",
              "xj352vofupe1dqz9emx13r": "optionTwo"
          },
          "questions": [
              "loxhs1bqm25b708cmbf3g",
              "vthrdm985a262al8qx3do"
          ]
      },
      "mtsamis": {
          "id": "mtsamis",
          "password": "xyz123",
          "name": "Mike Tsamis",
          "avatarURL": null,
          "answers": {
              "xj352vofupe1dqz9emx13r": "optionOne",
              "vthrdm985a262al8qx3do": "optionTwo",
              "6ni6ok3ym7mf1p33lnez": "optionOne"
          },
          "questions": [
              "6ni6ok3ym7mf1p33lnez",
              "xj352vofupe1dqz9emx13r"
          ]
      },
      "zoshikanlu": {
          "id": "zoshikanlu",
          "password": "pass246",
          "name": "Zenobia Oshikanlu",
          "avatarURL": null,
          "answers": {
              "xj352vofupe1dqz9emx13r": "optionOne"
          },
          "questions": []
      }
  },
  "polls": {
      "8xf0y6ziyjabvozdd253nd": {
          "id": "8xf0y6ziyjabvozdd253nd",
          "author": "sarahedo",
          "timestamp": 1467166872634,
          "optionOne": {
              "votes": [
                  "sarahedo"
              ],
              "text": "Build our new application with Javascript"
          },
          "optionTwo": {
              "votes": [],
              "text": "Build our new application with Typescript"
          }
      },
      "6ni6ok3ym7mf1p33lnez": {
          "id": "6ni6ok3ym7mf1p33lnez",
          "author": "mtsamis",
          "timestamp": 1468479767190,
          "optionOne": {
              "votes": [],
              "text": "hire more frontend developers"
          },
          "optionTwo": {
              "votes": [
                  "mtsamis",
                  "sarahedo"
              ],
              "text": "hire more backend developers"
          }
      },
      "am8ehyc8byjqgar0jgpub9": {
          "id": "am8ehyc8byjqgar0jgpub9",
          "author": "sarahedo",
          "timestamp": 1488579767190,
          "optionOne": {
              "votes": [],
              "text": "conduct a release retrospective 1 week after a release"
          },
          "optionTwo": {
              "votes": [
                  "sarahedo"
              ],
              "text": "conduct release retrospectives quarterly"
          }
      },
      "loxhs1bqm25b708cmbf3g": {
          "id": "loxhs1bqm25b708cmbf3g",
          "author": "tylermcginnis",
          "timestamp": 1482579767190,
          "optionOne": {
              "votes": [],
              "text": "have code reviews conducted by peers"
          },
          "optionTwo": {
              "votes": [
                  "sarahedo"
              ],
              "text": "have code reviews conducted by managers"
          }
      },
      "vthrdm985a262al8qx3do": {
          "id": "vthrdm985a262al8qx3do",
          "author": "tylermcginnis",
          "timestamp": 1489579767190,
          "optionOne": {
              "votes": [
                  "tylermcginnis"
              ],
              "text": "take a course on ReactJS"
          },
          "optionTwo": {
              "votes": [
                  "mtsamis"
              ],
              "text": "take a course on unit testing with Jest"
          }
      },
      "xj352vofupe1dqz9emx13r": {
          "id": "xj352vofupe1dqz9emx13r",
          "author": "mtsamis",
          "timestamp": 1493579767190,
          "optionOne": {
              "votes": [
                  "mtsamis",
                  "zoshikanlu"
              ],
              "text": "deploy to production once every two weeks"
          },
          "optionTwo": {
              "votes": [
                  "tylermcginnis"
              ],
              "text": "deploy to production once every month"
          }
      }
  },
  "loadingBarReducer": {
      "default": 3
  }
};
