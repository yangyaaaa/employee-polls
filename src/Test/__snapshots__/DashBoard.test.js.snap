// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashBoard will catch a snapshot for DashBoard 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="poll-list-container"
      >
        <div>
          <button
            data-testid="answered-polls-test-id"
          >
            Answered Polls
          </button>
          <button
            data-testid="unanswered-polls-test-id"
          >
            UnAnswered Polls
          </button>
        </div>
        <div>
          <ul>
            <li>
              <a
                class="poll-item"
                href="/questions/xj352vofupe1dqz9emx13r"
              >
                <h4>
                  Would you rather deploy to production once every two weeks or deploy to production once every month?
                   
                </h4>
                <div
                  class="poll-metadata"
                >
                  <span
                    class="author-name"
                  >
                    created by: 
                    Mike Tsamis
                  </span>
                  <span
                    class="poll-date"
                  >
                    creation Date: 
                    9:16:PM | 4/30/2017
                  </span>
                </div>
              </a>
            </li>
            <li>
              <a
                class="poll-item"
                href="/questions/vthrdm985a262al8qx3do"
              >
                <h4>
                  Would you rather take a course on ReactJS or take a course on unit testing with Jest?
                   
                </h4>
                <div
                  class="poll-metadata"
                >
                  <span
                    class="author-name"
                  >
                    created by: 
                    Tyler McGinnis
                  </span>
                  <span
                    class="poll-date"
                  >
                    creation Date: 
                    1:09:PM | 3/15/2017
                  </span>
                </div>
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="poll-list-container"
    >
      <div>
        <button
          data-testid="answered-polls-test-id"
        >
          Answered Polls
        </button>
        <button
          data-testid="unanswered-polls-test-id"
        >
          UnAnswered Polls
        </button>
      </div>
      <div>
        <ul>
          <li>
            <a
              class="poll-item"
              href="/questions/xj352vofupe1dqz9emx13r"
            >
              <h4>
                Would you rather deploy to production once every two weeks or deploy to production once every month?
                 
              </h4>
              <div
                class="poll-metadata"
              >
                <span
                  class="author-name"
                >
                  created by: 
                  Mike Tsamis
                </span>
                <span
                  class="poll-date"
                >
                  creation Date: 
                  9:16:PM | 4/30/2017
                </span>
              </div>
            </a>
          </li>
          <li>
            <a
              class="poll-item"
              href="/questions/vthrdm985a262al8qx3do"
            >
              <h4>
                Would you rather take a course on ReactJS or take a course on unit testing with Jest?
                 
              </h4>
              <div
                class="poll-metadata"
              >
                <span
                  class="author-name"
                >
                  created by: 
                  Tyler McGinnis
                </span>
                <span
                  class="poll-date"
                >
                  creation Date: 
                  1:09:PM | 3/15/2017
                </span>
              </div>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
